waiter/docs/service-references.md

```diff
@@ -0,0 +1,78 @@
+# Service References
+
+To process a http request, Waiter needs to resolve a request to a service description.
+This service description tells Waiter how to locate the service instance to use to process the current request.
+
+Waiter uses request headers to construct the service description using the `ServiceDescriptionBuilder`.
+It is possible for the service to be reachable from various combinations of headers:
+  1. token - single or multiple
+  1. on-the-fly services without tokens
+  1. on-the-fly services that extend tokens, e.g. run-as-requester
+
+## Service GC
+
+By default, Waiter relies on the service idle timeout to GC services after periods of inactivity (not receiving requests).
+However, services reachable only via references, e.g. tokens, can be GC-ed eagerly if the reference has been updated.
+
+When a service description is constructed from a request, the service references are also updated.
+These references are available as the `:reference-type->entry` key in the descriptor.
+The `reference-type->entry` is a map where the keys represent a `:type` parameter.
+
+### Reference (`:reference-type->entry` in the descriptor) examples
+
+- A service created with on-the-fly header without tokens will have an empty map
+  as the following value for `:reference-type->entry` in the descriptor:
+```
+  {}
+```
+
+- A service created with `x-waiter-token: foo` header on a request will have the
+  following value for `:reference-type->entry` in the descriptor:
+```
+  {:token {:sources [{:token "foo" :version "v0"}]}}
+```
+
+- A service created with `x-waiter-token: bar,baz` header on a request will have the
+  following value for `:reference-type->entry` in the descriptor:
+```
+  {:token {:sources [{:token "bar" :version "v1"}
+                     {:token "baz" :version "v2"}]}}
+```
+
+If the same service can be accessed by another token, we end up building multiple
+  `reference-type->entry` maps as references that refer to the service.
+E.g. if all the above requests mapped to the same service, that service would have
+  the following references:
+```
+  #{{}
```

Comment:
Is `:reference-type->entry` a set or a map?  In previous examples it was a map, but now it's a set.

Reply:
`:reference-type->entry` is a map computed per request. Over the lifetime of the services individual `:reference-type->entry` are aggregated into a set to maintain all references that have been used to access the service.